include/cuco/detail/hash_functions/sigbithash.cuh

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/hash_functions/utils.cuh>
+#include <cuco/extent.cuh>
+
+#include <cstddef>
+#include <cstdint>
+
+namespace cuco::detail {
+
+/**
+ * @brief An SigBit hash function to hash the given argument on host and device.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * @tparam Key The type of the values to hash
+ */
+template <typename Key>
+struct SigBitHash_32 {
+  using argument_type = Key;            ///< The type of the values taken as argument
+  using result_type   = std::uint32_t;  ///< The type of the hash values produced
+
+  /**
+   * @brief Constructs a SigBitHash_32 hash function by finding the significant bits of
+   * a given span of keys
+   *
+   * @param start 	A pointer/iterator to the first key in the list
+   * @param numKeys 	The number of keys in the span
+   */
+  __host__ __device__ constexpr SigBitHash_32(Key* start, std::uint32_t numKeys)
```

Comment:
question: is there a seed parameter for sighash?

Reply:
No.  And it doesn't need one, actually!

The point of a seed in a traditional hash is so that a hash will not behave the same each time.  This is to prevent denial-of-service attacks based on hash collisions, like this: https://stackoverflow.com/questions/52184366/why-does-hashmap-need-a-cryptographically-secure-hashing-function

However, a perfect hash has no collisions.  (Collision is defined as `hash(x) == hash(y) && x != y`.)  So with no collisions, there is no need for the seed.

Reply:
OK, got it. In that sense, we need to figure out a way to guarantee that the sig hash is used exclusively with perfect probing.

Reply:
Well, kind of.  To be more accurate, we need to make sure that a perfect probing scheme is **always** used with a hasher that is perfect.  For example, use **can** use `SigBitHash` with perfect probing but you can also use it with double probing or linear probing, no problem!  What we want to avoid is someone trying to use a non-perfect hash together with perfect probing.  Like perfect probing with xxhash.  That's not allowed, it will give invalid results or crash!

I'll try to write up the whole idea here:

# Background

A hash table, when constructed, gets a probing scheme that is a [functor](https://en.wikipedia.org/wiki/Function_object).  For example, linear or double or perfect, etc.  The probing scheme describes how the hasher will scan through the slots looking for a place to insert/lookup.

That probing scheme has inside it a hasher which is also a functor.  That hasher might be xxhash or murmur3 or Identity or SigBitHash, etc.  The hasher does the math to help the probing scheme get the slot to look at.

If:

* the hasher is perfect (`hash(x) == hash(y) => x==y`) **and**
* the maximum output of the hasher's `operator()` is less than the number of slots in the hash table **and** 
* the minimum output of the hasher's `operator()` is at least 0

then the hashtable or hashset can:

* Skip the use of atomics in most cases (because all collisions will just be the same value).
* Run a faster equality operator (because all collisions are sure to be equal).
* Skip the modulo operator (because the hasher's output is always less than the slot count).

These will give a speed improvement.

# Goal

We want to make it as easy as possible for the programmer using cuco to get this right and not screw it up.

# Solutions

## Detecting perfect probing scheme.

One way is that we could accomplish this is to have a prober class that is called `perfect_probing`, just like we already have one called `linear_probing`.  And in the code for the hash table, in the places where we use modulo or check equality or use atomics, we just do a little `if constexpr` to check if the type of the probing class is `perfect_probing`.  And if it is, then we skip those and do things the fast way.

The problem with this is that, what if the user wants to invent their own perfect probing scheme?  Because we are testing against a specific class `perfect_probing` then the user won't be able to create a class called `my_perfect_probing` and use it as if it is perfect.  This is not good.

So instead, we put a method inside the `perfect_probing` class that is called `is_perfect` that returns `true` only for perfect probing schemes.  And the `if constexpr` above is now testing that the probing scheme's `is_perfect` method returns `true`.

```c++
if constexpr (this.probing_scheme_.is_perfect()) {
  // don't use atomics
} else {
  // do it the normal way with atomics.
}
```

### Detecting if a class has a method.

The problem with this is that the other probing schemes don't even have a function called `is_perfect` so they will fail to compile.

```c++
 // If this.probing_scheme_ is linear probing, compile-time error.
if constexpr (this.probing_scheme_.is_perfect()) {
  // don't use atomics
} else {
  // do it the normal way with atomics.
}
```


But we can solve this using an idea similar to [c++ type traits](https://accu.org/journals/overload/9/43/frogley_442/).  We can use [SFINAE](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error) to [detect if a class has a method at compile time](https://stackoverflow.com/questions/87372/check-if-a-class-has-a-member-function-of-a-given-signature).  It is done [here](https://github.com/NVIDIA/cuCollections/pull/505/files#diff-2d1ed9c31db1250809c8d1f33971551ab468bc7521f037c2e61291848158ad1aR184-R192).

So the test is now, if:
* the hash table's probing scheme has the method `is_perfect` **and**
* the `is_perfect` method returns `true`

then:
* do all the perfect hashing optimization.

else:
* do regular hash table stuff.

This will allow anyone to make a new perfect probing scheme just by defining `is_perfect()` to return `true`.  It also means that we *don't* have to put an `is_perfect` method into all the other probing schemes because they all get treated as if their `is_perfect` returns `false`.

We're going to need to do this a few times so we make a utility that can check if a class has a method and, if so, it calls that method and returns the result.  If it *doesn't* have the method then it returns a default value that we provide.   A code example is below and you can see it [on godbolt here](https://godbolt.org/z/5nnYM3ccW).

<details>
  <summary>Code to call a class method and use a default if the method doesn't exist.</summary>

```c++
#include <stdio.h>

/*** The utility code ***/
template <typename T, typename = void>
struct has_foo_method_helper : std::false_type {};

template <typename T>
struct has_foo_method_helper<T, std::void_t<decltype(std::declval<T>().foo())>> : std::true_type {};;

template <typename T>
bool constexpr has_foo_method() {
    return has_foo_method_helper<T>::value;
}

template <typename T>
bool constexpr call_foo_with_default(bool default_value) {
    if constexpr (has_foo_method<T>()) {
        return T::foo();
    }
    return default_value;
}

/*** Example ***/

struct HasFooTrue {
    static bool constexpr foo() { return true; }
};

struct HasFooFalse {
    static bool constexpr foo() { return false; }
};

struct DoesNotHaveFoo {
};

int main() {
  printf("%d\n", call_foo_with_default<HasFooTrue>(false));
  printf("%d\n", call_foo_with_default<HasFooFalse>(false));
  printf("%d\n", call_foo_with_default<DoesNotHaveFoo>(false));

  printf("%d\n", call_foo_with_default<HasFooTrue>(true));
  printf("%d\n", call_foo_with_default<HasFooFalse>(true));
  printf("%d\n", call_foo_with_default<DoesNotHaveFoo>(true));

  return 0;
}
```
</details>

We can do this for any method.  We'll call functions using this technique `call_XXXX_with_default` and `has_XXXX_method`.

The hash table, to check if we can skip atomics or modulo or equality comparison, will use:

```c++
if constexpr (call_is_perfect_with_default<probing_scheme>(false)) {
  // it is perfect!  don't use atomics, modulo, etc
} else {
  // not perfect hashing so do it the normal way, with atomics, modulo, etc
}
```

## Checking that the hasher is perfect.

With the above done, now we can write code that will work one way for perfect probing and a different way for linear and double probing.  And we don't need to modify our other probing schemes.  If a user wants to make a new probing scheme, they can make it the normal way and it will not be perfect.  Or they can define `is_perfect` to return `true` and it will use the optimizations of perfect hashing.  It's all pretty good.

The next thing that we need to check is that the hasher provided to the perfect probing scheme is indeed perfect.  If the user constructs a `perfect_probing` instance and supplies to it `xxhash`, which isn't perfect, it'll fail!

One solution to this would be to just have a comment in the docs that says that if you use a `perfect_probing`, make sure you also use a perfect hasher functor.  But we can do better.

What we'll do is put a method in the perfect hashers (such as `SigbitHash` and `IdentityHash`) called `is_perfect` that returns `true`.  And in the constructor of `perfect_probing`, we will have:

```c++
static_assert(call_is_perfect_with_default<Hasher>(false));
```

With this, if the user accidentally tries to construct a `perfect_probing` with a hasher that is not perfect, it'll be a compile time error.

## Checking that the range of the hasher is less than the number of slots.

With the techniques above, we are now ensuring that only when the probing is perfect do we skip atomics/modulo/etc.  And the perfect probing always has a perfect hasher.  The last thing to check is that the max output of the hasher is less than the number of slots.  In mathematics, the inputs to a function is called its "domain" and the output is the "range".  So we'll say that we want the `range_max` of the hasher to be less than number of slots in the hash table.

For this, the hasher will have a method called `range_max` that returns that maximum possible value of the output of the hasher's `operator()`.  For hashers that don't have the `range_max` method, we'll just assume that the maximum is the maximum of the output type of `operator()`.  We can get that maximum using [code like this on godbolt](https://godbolt.org/z/G3qvjjq9z).

<details>
  <summary>Code to get the return type max of `operator()`.</summary>

```c++
#include <stdio.h>
#include <type_traits>
#include <limits>

struct FooReturnsInt {
    int operator()() { return 0; }
};

struct FooReturnsLong {
    long operator()() { return 0; }
};

int main() {
  printf("%lld\n", std::numeric_limits<std::invoke_result_t<FooReturnsLong>>::max());
  printf("%lld\n", std::numeric_limits<std::invoke_result_t<FooReturnsInt>>::max());

  return 0;
}
```
</details>

Let's call that `get_invoke_result_t_max<T>()`.  So now we will add to the constructor of our hashtable code like this:

```c++
if constexpr (call_is_perfect_with_default<probing_scheme>(false)) {
  // Okay, this is a perfect probing scheme.  Is the range of the hasher valid?
  constexpr auto max_of_output_type = get_invoke_result_t_max<probing_scheme>();
  if (call_range_max_with_default<probing_scheme>(max_of_output_type) >= slots_.size()) {
    // Uh oh, error, the perfect probing scheme is valid
    // but this hash table doesn't have enough slots to use it!
  }
}
```

And the `probing_scheme::range_max` function will call its hasher's `range_max` function.  

With this, now we can check at the time of constructing the hash table that the probing scheme provided will indeed use a hasher that does not return values greater than the number of the slots.  This lets us get rid of the modulo operator and that is necessary for removing atomics, too.

All the above should also be done for `range_min`.  (However, `range_min` is usually not going to be defined because the `operator()` functions will usually return an unsigned value and the `call_range_min_with_default` will already be returning a default of 0, which is correct.  So we need the test but the hasher probably doesn't need to define `range_min` for unsigned outputs.)

## Reporting a range_max error.

We won't know the `range_max` of a hasher at compile time so we can only do the `call_range_max_with_default` at runtime, and the `range_max` functions are not `constexpr`, either, for the same reason.

The test above happens inside the constructor of the hash table.  Constructors have only one way to return errors, by throwing exceptions.  It's [perfectly acceptable to throw an exception in a constructor.  ](https://isocpp.org/wiki/faq/exceptions#ctors-can-throw).  We can use the existing `CUCO_FAIL` macro for this.

```c++
if constexpr (call_is_perfect_with_default<probing_scheme>(false)) {
  // Okay, this is a perfect probing scheme.  Is the range of the hasher valid?
  constexpr auto max_of_output_type = get_invoke_result_t_max<probing_scheme>();
  auto max_hasher_output =
      call_range_max_with_default<probing_scheme>(max_of_output_type);
  if (max_hasher_output >= slots_.size()) {
    std::stringstream error_string;
    error_string << "Can't use perfect probing because the number of slots is only "
                 << slots_size()
                 << " but the hasher may return the value "
                 << max_hasher_output
                 << " which was determined because ";
    if (has_range_max<hasher>()) {
      error_string << "we called range_max and that's what it told us"
    } else {
      error_string << "there's no range_max method so we assumed the maximum of the data type"
    }
     CUCO_FAIL(error_string.str(), std::runtime_error);
  }
}
```

(The exception message above is extra verbose just to make clear the logic.)

As before, do the same for `range_min`.

## Avoiding the runtime error.

In practice, a smart user of cuco's perfect hashing is never going to see this exception because they are going to first construct the hasher and then look at the range_max of it and use that info to determine if they want to use perfect hashing or not.  In psuedocode, here's what the user will do:

```c++
vector <int> data_to_insert;
SigBitHash my_hasher(data_to_insert); // The SigBitHash looks at the data to make a perfect hash.
auto range_max = my_hasher.range_max();
constexpr auto max_hash_memory = 1000000; // Based on how much memory we want to allow our hash table to use.
if (range_max < max_hash_memory) {
  // Cool!  We can use perfect hashing!
  auto my_probing_scheme = perfect_probing(my_hasher);
  // remember to add one because of fencepost error https://en.wikipedia.org/wiki/Off-by-one_error
  auto my_hash_table = hash_table(my_probing_scheme, range_max + 1);
  // use the hash table
} else {
  // Okay, we'll just use regular hashing.
  auto my_probing_scheme = linear_probing(my_hasher);
  auto occupancy = 2; // We'll have twice as many slots as item to insert.
  auto my_hash_table = hash_table(my_probing_scheme, data_to_insert.size() * occupancy);
  // use the hash table
}
```

To make it ever tidier, you can put the above into a function.  Instead of using the hash table, just return the hash table.  The return type of the function will have to be `std::variant<...>` where the variants are either a hash table with perfect probing or a hash table with linear probing.  And then you can use the `std::visit` with a [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) to do `insert` and `lookup` on the hash table.

Or if you don't like the visitor pattern, just make a function called `make_perfect_hash_table` that has return type `std::optional<...>` that returns a perfect hash table when `range_max < max_hash_memory` and otherwise returns `std:nullopt`.  And then call `make_perfect_hash_table`.  If it returns a hash table, use it.  If not, fall back to using linear probing.

Both of these solutions will result in the same compiled asm code but it's up to the user which one seems more ergonomic.  Doesn't really matter, they work the same.

# Conclusion

With all of this, we'll have a `perfect_probing` that gets us all the optimizations we want and provides a lot of safety so that the user won't use perfect_probing incorrectly.  None of the existing probing schemes and hash functions need to be modified and if the user doesn't want to use perfect probing, they don't have to change anything!

If the user *does* want to use perfect probing, they can.  It will do all the compile-time and run-time checks to make sure that that the hasher is valid to use and the hash table size is correct.  The only overhead will be a single test for `range_max < slots_.size()`, once, on the CPU, at construction time.  It's trivially fast.  That test throws an exception but they can avoid it by doing the test themselves before construction of the hash table.  In fact, it'll be the same test so the optimizer will optimize it away and remove internal overhead in the constructor!

If the user wants to design their own hasher that isn't perfect, they can do it like they are already doing it.

If they want to design a perfect one, they just have to remember to put the `is_perfect` method in there that returns `true`.  And if they forget to do that but try to use it with `perfect_probing`, they will get a compile-time error that explains the problem.

If they forget to define `range_max` in their perfect hasher then the default will be the maximum of the output type.  If that's too big then they will get a runtime error explaining the that the range is too big for the number of slots and they can fix that.  

## Possibilities for errors

It's pretty foolproof.

One way to break it would be if the user defines a hasher that isn't perfect and they put in it a method called `is_perfect` that returns `true`.  I mean, that would be pretty unlikely to do on accident!

Another way to break it is if they define the `range_max` function of the hasher and then return a max is the smaller than what the hasher's `operator()` sometimes returns.  This is only possible if the user's hash function is broken.

To prevent these, we'll have comments in the code around these functions that explain their purpose so that a user who is designing a perfect hasher will know how to write the above functions correctly.

# Notes

* Perfect hashing is only possible for hashset and hashtable.  It won't work for multihash nor multiset.

Reply:
To simplify the problem, let's narrow down the discussion to some key points:

- From cuco's perspective, we don't allow user-defined probing schemes
- Whether it's perfect or not, that's the information carried by probing scheme, not hasher
- The steps checking perfect hasher, checking the range, reporting range max error and avoiding runtime errors should be the responsibility of a helper function like `create_perfect_hashing` whose return type is an optional: if the input keys can be hashed perfectly, returns the perfect probing. Otherwise, returns `nullopt`.  
- `perfect_probing` ctor should be private to avoid misuse and the only way to get a perfect probing is via the above helper function. By doing so, we can avoid all the potential errors mentioned in your `Possibilities for errors` section